proxy/http/HttpCacheSM.cc

```diff
@@ -283,6 +283,7 @@ HttpCacheSM::do_cache_open_read(const HttpCacheKey &key)
     return ACTION_RESULT_DONE;
   } else {
     ink_assert(pending_action != nullptr || write_locked == true);
+    captive_action.cancelled = false; // Make sure not cancelled before we hand it out
     return &captive_action;
```

Comment:
Not about this change, but I'm not sure why this function return `&captive_action` instead of `pending_action`?

Reply:
No, the captive action is the fixed storage for the action that the HttpCacheSM functions return.  

If you look in HttpSM, you'll see 
```
  pending_action =
    c_sm->open_write(&key, s_url, &t_state.hdr_info.client_request, object_read_info,
                     static_cast<time_t>((t_state.cache_control.pin_in_cache_for < 0) ? 0 : t_state.cache_control.pin_in_cache_for),
                     retry, allow_multiple);
```
So the HttpSM's pending_action is pointing at the HttpCacheSM captive_action.  If for some reason, the HttpSM no longer wants to deal with the HttpCacheSM result, it can cancel the captive_action that it's pending action is pointing at.  

The HttpCacheSM's pending action is returned from the scheduling subsystem and may even be null.  So when the HttpCacheSM is cleaning up, it should be canceling the action it's pending_action is pointing at.

Reply:
OK, HttpCacheSM's captive_action is an action for handling HttpCacheSM from HttpSM. Well, when HttpSM cancels HttpCacheSM's captive_action, what HttpCacheSM should do?

Reply:
> The captive_action::cancel() passes through to cancel the pending_action member of the HttpCacheSM.
https://github.com/apache/trafficserver/pull/7819#issuecomment-839860155

I missed this part. I agree we can't get rid of `captive_action`.